akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala

```diff
@@ -257,6 +316,106 @@ private[akka] object EventWriter {
               Behaviors.same
           }
 
+        def handleWrite(repr: PersistentRepr, replyTo: ActorRef[StatusReply[WriteAck]]): Unit = {
+          val persistenceId = repr.persistenceId
+          val sequenceNumber = repr.sequenceNr
+          val newStateForPid =
+            perPidWriteState.get(persistenceId) match {
+              case None =>
+                if (fillSequenceNumberGaps && sequenceNumber != 1L) {
+                  askMaxSeqNr(persistenceId, originalErrorDesc = None)
+                  StateForPid(
+                    waitingForReply = Map.empty,
+                    waitingForSeqNrLookup = Vector((repr, replyTo)),
+                    currentTransactionId = 1)
+                } else {
+                  sendToJournal(1, Vector(repr))
+                  StateForPid(Map((repr.sequenceNr, (repr, replyTo))), currentTransactionId = 1)
+                }
+
+              case Some(state) =>
+                val expectedSeqNr =
+                  if (fillSequenceNumberGaps) state.nextExpectedSeqNr
+                  else repr.sequenceNr
+
+                if (state.seqNrlookupInProgress) {
+                  context.log.trace2(
+                    "Seq nr lookup in progress for persistence id [{}], adding sequence nr [{}] to pending",
+                    persistenceId,
+                    sequenceNumber)
+                  state.copy(waitingForSeqNrLookup = state.waitingForSeqNrLookup :+ ((repr, replyTo)))
+                } else if (fillSequenceNumberGaps && state.latestSeqNr == 0L && state.idle && sequenceNumber != 1L) {
```

Comment:
I wrote some tests of the actual filling, and it works out quite nicely. In many cases we don't have to lookup the max when we have already a confirmed seqNr or we have some pending that is supposed to be successful and after that become a confirmed latest seqNr. I'll have to add some tests of the error scenarios for that though.